File: Streaming Data(Bloom Filter, Flajolet-Martin Algorithm, Fixed Size Sampling)/python/task3_Fixed_Size_Sampling.py
from blackbox import BlackBox
import sys
import time

import random

random.seed(553)

if __name__ == '__main__':

    input_filename = sys.argv[1]
    stream_size = int(sys.argv[2])
    num_of_asks = int(sys.argv[3])
    output_filename = sys.argv[4]


    random.seed(553)
    t = time.time()
    mem = []
    bx = BlackBox()
    count = 0
    with open(output_filename, 'w') as file:
        file.write("seqnum,0_id,20_id,40_id,80_id\n")
        for _ in range(num_of_asks):
            stream_users = bx.ask(input_filename, stream_size)
            if stream_size * (_ + 1) <= 100:
                print("Case 1")
                for idx, user in enumerate(stream_users):
                    mem.append(user)
                    count += 1

            if stream_size * (_ + 1) > 100:
                idx = 0
                while len(mem) < 100:
                    print("Case 2")
                    mem.append(stream_users[idx])
                    idx += 1
                    count += 1

                while idx < stream_size:
                    count += 1
                    a = random.random()
                    if a > 100/count:
                        pass
                    elif a <= 100/count:
                        b = random.randint(0, 99)
                        mem[b] = stream_users[idx]

                    idx += 1
            seqnum = stream_size * (_ + 1)
            id_0 = mem[0]
            id_20 = mem[20]
            id_40 = mem[40]
            id_60 = mem[60]
            id_80 = mem[80]
            file.write(str(seqnum) + "," + id_0 + "," + id_20 + "," + id_40 + "," + id_60 + "," + id_80 + '\n')

    file.close

    print("Duration : ", time.time() - t)

File: 5. Streaming Data(Bloom Filter, Flajolet-Martin Algorithm, Fixed Size Sampling)/scala/task3_Fixed_Size_Sampling.scala
import java.io.{BufferedWriter, File, FileWriter}
import scala.io.Source

class Blackbox{
  private val r1 = scala.util.Random
  r1.setSeed(553)
  def ask(filename: String, num: Int): Array[String] = {
    val input_file_path = filename

    val lines = Source.fromFile(input_file_path).getLines().toArray
    val stream = new Array[String](num)

    for (i <- 0 to num-1){
      stream(i) = lines(r1.nextInt(lines.length))
    }
    return stream
  }
}

object task3 {

  def main(args: Array[String]): Unit = {
    val r1 = scala.util.Random
    r1.setSeed(553)

    val t1 = System.nanoTime

    // get argument
    val input_filename = args(0)
    val stream_size = args(1).toInt
    val num_of_asks = args(2).toInt
    val output_filename = args(3)

    var mem : List[String] = List()
    val box = new Blackbox
    var count = 0
    val random_object = scala.util.Random
    val file = new File(output_filename)
    val bw = new BufferedWriter(new FileWriter(file))
    bw.write("seqnum,0_id,20_id,40_id,80_id\n")

    var a : Float = 0
    var b : Int = 0

    for ( k <- 0 to num_of_asks-1){
      var stream_users : List[String] = box.ask(input_filename, stream_size).toList
      if (stream_size * ( k + 1) <= 100) {
        for ((user, idx) <- stream_users.zipWithIndex) {
          mem = mem ++ List(user)
          count = count + 1
        }
      }

      if (stream_size * (k + 1) > 100){
        var idx = 0
        while (mem.size <100){
          mem = mem ++ List(stream_users(idx))
          idx  = idx + 1
        }

        while (idx < stream_size){
          count = count + 1

          a = random_object.nextFloat()

          if (a <= 100/(count.toFloat)){
            b = random_object.nextInt(100)
            mem = mem.updated(b, stream_users(idx))
          }
          idx = idx + 1
        }
      }
      var seqnum = stream_size * (k + 1)
      var id_0 = mem(0)
      var id_20 = mem(20)
      var id_40 = mem(40)
      var id_60 = mem(60)
      var id_80 = mem(80)
      bw.write(seqnum.toString)
      bw.write(",")
      bw.write(id_0)
      bw.write(",")
      bw.write(id_20)
      bw.write(",")
      bw.write(id_40)
      bw.write(",")
      bw.write(id_60)
      bw.write(",")
      bw.write(id_80)
      bw.write("\n")

    }
    bw.close()

    val duration = ((System.nanoTime - t1) / 1e9d).toString
    println("Duration: ".concat(duration))






  }
}

